### PR TITLE
#237: Implement memory-aware criterion using user-defined memory constraints and update LBAF-Viz to keep track of these

### LINKED ISSUE #237: Implement memory-aware criterion using user-defined memory constraints in JSON
```
	"user_defined": {
						"block_bytes": 5669120.0,
						"block_id": 12,
						"task_bytes": 4617048.0
					}
	
```

### PR DESCRIPTION
**Pull request by ppebay**
Resolves #237

### CODE REVIEW

**Comment by ppebay** [comment]
As part of this PR the LBAF-Viz is improved, not showing above-constraint values (which happen for instance when the max memory upper-bound is **not** satisfied):
Screen Shot 2022-11-28 at 1 00 07 AM

**Comment by ppebay** [comment]
Visualizer was modified as part of this work: the per-rank QOI can now be changed at will:

### **With rank work:**

Screen Shot 2022-11-29 at 5 15 33 PM

Screen Shot 2022-11-29 at 5 15 18 PM

### **With rank load:**

Screen Shot 2022-11-29 at 5 19 29 PM

Screen Shot 2022-11-29 at 5 19 41 PM

**Comment by ppebay** [comment]
One PR #315 is merged, we will be able to view the memory-related QOIs @lifflander @nlslatt @nmm0 @PhilMiller

**Comment by ppebay** [comment]
Now tracking/visualizing arbitrary rank-level QOI, i.e., `max_memory_usage`:

Screen Shot 2022-11-29 at 11 58 41 PM
Screen Shot 2022-11-29 at 11 58 53 PM

**Comment by ppebay** [comment]
Comment for reviewers: constrained optimization is **not** part of this PR. A subsequent issue will deal with it.

File: src/lbaf/IO/lbsMeshBasedVisualizer.py [hunk 1]
```diff
@@ -316,8 +357,9 @@ def create_object_mesh(self, phase: Phase, object_mapping: set):
         # Assign edge volume values
         self.__logger.debug(f"\tedges:")
         for e in range(n_e):
-            v_arr.SetTuple1(e, edges.get(index_to_edge[e], float("nan")))
-            self.__logger.debug(f"\t{e} {index_to_edge[e]}): {v_arr.GetTuple1(e)}")
+            i2e =index_to_edge[e]
```
Comment (nlslatt) [inline src/lbaf/IO/lbsMeshBasedVisualizer.py:360]: Do you want to fix the spacing so that it doesn't end up a diff on the next PR? ;-)

**Review by nlslatt — COMMENTED** [review]
Looks reasonable to me, but I recommend a second person take a look.

File: src/lbaf/Model/lbsAffineCombinationWorkModel.py [hunk 2]
```diff
@@ -10,35 +11,53 @@ class AffineCombinationWorkModel(WorkModelBase):
...
-        """
-        # Compute affine combination of load and volumes
-        return self.__alpha * rank.get_load() + self.__beta * max(
+            alpha * load + beta * max(sent, received) + gamma,
+            under optional strict upper bounds."""
+
+        # Check whether strict bounds are satisfied
+        for k, v in self.__upper_bounds.items():
+            if getattr(rank, f"get_{k}")() > v:
+                return math.inf
```
Comment (nlslatt) [inline src/lbaf/Model/lbsAffineCombinationWorkModel.py:46]: Is this the problem?

**Comment by ppebay** [comment]
Bug that occurred when **no** default maximum for visualization QOI was specified, is now fixed:

Screen Shot 2022-11-30 at 11 35 03 PM
Screen Shot 2022-11-30 at 11 35 11 PM
Screen Shot 2022-11-30 at 11 35 20 PM

**N.B.:** in this case now, the space-time max of the QOI will be used.

**Comment by ppebay** [comment]
Also here is the behavior when an upper bound **was specified** (for the optimization process) for the **same** QOI as the one selected to be visualized:

Screen Shot 2022-11-30 at 11 38 08 PM


Screen Shot 2022-11-30 at 11 37 52 PM

**Comment by ppebay** [comment]
@nlslatt I think all comments are addressed

**Review by nlslatt — APPROVED** [review]
Looks good AFAICT

**Pull request merged**